pkg/backup/backupsink/sst_sink_key_writer.go

```
@@ -0,0 +1,242 @@
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	lastFile, err := s.maybeDoSizeFlush(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span. It will either open a new
+// BackupManifest_File, or if the new span contiguously extends the last span,
+// it will extend it if it remains under the fileSpanByteLimit. The caller is
+// responsible for ensuring that spans that are reset on do not end mid-row.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	// At this point, we can assume the new span is either contiguous with the
+	// last span or comes after the last span (if it exists).
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
```

why call `setMidRowForPrevKey` in 3 different places? Couldn't we just call it in `WriteKey` after we validate the key? If we do that, then `Flush()` and `Reset()` should return an error if s.Midkey is true and maybeFlushOnSize continues to noop.

---

We could call it in `WriteKey` twice and remove the need to call it again in the other functions.

Let's say I have the following keys I am writing in the span `[a, d)`:

1. `a`
2. `b/1`
3. `b/2`
4. `c`

Upon calling `WriteKey` on `b/1`, I can use `b1` to validate that `a` is not mid-row. But until I write `b/2`, I don't know that `b/1` is mid-row, so the best guess I can make is using the end-key of `[a, d)` to determine if `b/1` is mid-row, which it would determine as not being mid-row. That means that when I'm writing `b/2`, I need to call `setMidRowForPrevKey` again on `b/1` before doing any of the flushing. So `WriteKey` would look like this:

```golang
func (s *SSTSinkKeyWriter) WriteKey(ctx, key, val) {
    s.validateKey(key)
    s.setMidRowForPrevKey(ctx, key)
    s.maybeDoSizeFlush(ctx)
    s.sst.PutRawMVCC(key, val)
    s.prevKey = key
    s.setMidRowForPrevKey(ctx, s.span.EndKey)
}
```

---

Oh, that's a good example I hadn't thought through. What if we did the following: 
```
func (s *SSTSinkKeyWriter) WriteKey(ctx, key, val) {
   // same pseudo code you had
    s.validateKey(key)
    s.setMidRowForPrevKey(ctx, key)
    s.maybeDoSizeFlush(ctx)
    s.sst.PutRawMVCC(key, val)
    s.prevKey = key
   // my additional suggestion
    s.midRow = true
}
```
i.e. after writing the key, we always set `s.midkey=True`  (perhaps were rename `s.midkey` to `s.CouldBeMidKey`). Then, if we assume that `Reset()` is only ever called after a finished `restoreSpanEntry`, then Reset can safely flip `s.CouldBeMidKey=false`. Once Reset sets midkey to false, the Flush will proceed without error. 

Your proposed check of `s.setMidRowForPrevKey(ctx, s.span.EndKey)` could leave s.MidRow=false even if the prev key is actually mid key.